Index: vorbis.c
===================================================================
--- squeezelite/vorbis.c	(revision 1213)
+++ squeezelite/vorbis.c	(working copy)
@@ -320,14 +320,14 @@
 
 static bool load_vorbis() {
 #if !LINKALL
-	void *handle = dlopen(LIBVORBIS, RTLD_NOW);
+	void *handle = dlopen(LIBTREMOR, RTLD_NOW);
 	char *err;
-	bool tremor = false;
+	bool tremor = true;
 
 	if (!handle) {
-		handle = dlopen(LIBTREMOR, RTLD_NOW);
+		handle = dlopen(LIBVORBIS, RTLD_NOW);
 		if (handle) {
-			tremor = true;
+			tremor = false;
 		} else {
 			LOG_INFO("dlerror: %s", dlerror());
 			return false;
