diff --git a/codec/EndianPortable.c b/codec/EndianPortable.c
index 5a7d5b8..b8423c9 100644
--- a/codec/EndianPortable.c
+++ b/codec/EndianPortable.c
@@ -40,6 +40,8 @@
 #define TARGET_RT_LITTLE_ENDIAN 1
 #elif defined (TARGET_OS_WIN32)
 #define TARGET_RT_LITTLE_ENDIAN 1
+#elif defined (__arm__) || defined(__aarch64__)
+#define TARGET_RT_LITTLE_ENDIAN 1
 #endif
 
 uint16_t Swap16NtoB(uint16_t inUInt16)
