Index: main.c
===================================================================
--- main.c	(revision 499)
+++ main.c	(working copy)
@@ -49,6 +49,7 @@
 		   "  -a <b>:<p>:<f>:<m>\tSpecify ALSA params to open output device, b = buffer time in ms or size in bytes, p = period count or size in bytes, f sample format (16|24|24_3|32), m = use mmap (0|1)\n"
 #endif
 #if PORTAUDIO
+#define OSXPPC
 #if PA18API
 		   "  -a <frames>:<buffers>\tSpecify output target 4 byte frames per buffer, number of buffers\n"
 #elif OSX && !defined(OSXPPC)
@@ -56,6 +57,7 @@
 #else
 		   "  -a <l>\t\tSpecify Portaudio params to open output device, l = target latency in ms\n"
 #endif
+#undef OSXPPC
 #endif
 		   "  -a <f>\t\tSpecify sample format (16|24|32) of output file when using -o - to output samples to stdout (interleaved little endian only)\n"
 		   "  -b <stream>:<output>\tSpecify internal Stream and Output buffer sizes in Kbytes\n"
Index: output_pa.c
===================================================================
--- output_pa.c	(revision 499)
+++ output_pa.c	(working copy)
@@ -23,7 +23,7 @@
 #include "squeezelite.h"
 
 #if PORTAUDIO
-
+#define OSXPPC
 #include <portaudio.h>
 #if WIN
 #define snprintf _snprintf
