diff --git a/ffmpeg.c b/ffmpeg.c
index 43be7ab..e7ffa1a 100644
--- a/ffmpeg.c
+++ b/ffmpeg.c
@@ -274,7 +274,7 @@ static decode_state ff_decode(void) {
 		ff->mmsh_bytes_left = ff->mmsh_bytes_pad = ff->mmsh_packet_len = 0;
 
 		if (!ff->readbuf) {
-			ff->readbuf = AV(ff, malloc, READ_SIZE +  AV_INPUT_BUFFER_PADDING_SIZE);
+			ff->readbuf = AV(ff, malloc, READ_SIZE +  FF_INPUT_BUFFER_PADDING_SIZE);
 		}
 
 		avio = AVIO(ff, alloc_context, ff->readbuf, READ_SIZE, 0, NULL, _read_data, NULL, NULL);
